### PR TITLE
Upgrade CI to PyTorch 1.13

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Dropping PT 1.13 from RC and use is as latest

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Comment by Borda** [comment]
@kaushikb11, are we aware of this Bagua limitation?
```
#12 0.700 ERROR: Could not find a version that satisfies the requirement bagua-cuda116 (from versions: none)
#12 0.700 ERROR: No matching distribution found for bagua-cuda116
```

File: .github/checkgroup.yml [hunk 1]
```diff
@@ -25,27 +25,25 @@ subprojects:
       - ".actions/**"
     checks:
       # Note: updates here should be applied to the lightning_lite group
-      - "pl-cpu (macOS-11, pytorch, 3.8, 1.10)"
-      - "pl-cpu (macOS-11, pytorch, 3.9, 1.11)"
-      - "pl-cpu (macOS-11, pytorch, 3.10, 1.12)"
+      - "pl-cpu (macOS-11, pytorch, 3.8, 1.11)"
+      - "pl-cpu (macOS-11, pytorch, 3.9, 1.12)"
+      - "pl-cpu (macOS-11, pytorch, 3.10, 1.13)"
       - "pl-cpu (macOS-11, pytorch, 3.7, 1.9, oldest)"
```
Comment (otaj) [inline .github/checkgroup.yml:31]: where is 1.10?

File: .github/workflows/ci-pytorch-tests.yml [hunk 2]
```diff
@@ -40,32 +40,28 @@ jobs:
...
+          - {os: "macOS-11", pkg-name: "pytorch", python-version: "3.9", pytorch-version: "1.12"}
+          - {os: "ubuntu-20.04", pkg-name: "pytorch", python-version: "3.8", pytorch-version: "1.11"}
+          - {os: "ubuntu-20.04", pkg-name: "pytorch", python-version: "3.9", pytorch-version: "1.12"}
           - {os: "ubuntu-20.04", pkg-name: "pytorch", python-version: "3.10", pytorch-version: "1.12"}
-          - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "1.12"}
+          - {os: "windows-2022", pkg-name: "pytorch", python-version: "3.8", pytorch-version: "1.11"}
+          - {os: "windows-2022", pkg-name: "pytorch", python-version: "3.9", pytorch-version: "1.12"}
           - {os: "windows-2022", pkg-name: "pytorch", python-version: "3.10", pytorch-version: "1.12"}
-          - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "1.12"}
+          # only run PyTorch latest with Python latest
```
Comment (otaj) [inline .github/workflows/ci-pytorch-tests.yml:51]: Currently latest python is 3.11, so this comment and actual code is out of sync. It seems, `PyTorch` `1.13` support `3.11`

Comment (akihironitta) [reply .github/workflows/ci-pytorch-tests.yml:51]: Here's our tracking issue #14652 for Python 3.11 support

File: .github/checkgroup.yml (hunk 1 above)
Comment (Borda) [reply .github/checkgroup.yml:31]: we do not have full range for mac

Comment (akihironitta) [reply .github/checkgroup.yml:31]: Why not? We used to test all PyTorch versions, right?

Comment (Borda) [reply .github/checkgroup.yml:31]: Yes with conda so on Ubuntu as we still, right?

**Comment by Borda** [comment]
@carmocca could you pls check the groups here? :innocent:

File: .github/workflows/ci-pytorch-dockers.yml [hunk 3]
```diff
@@ -24,7 +24,8 @@ concurrency:
   cancel-in-progress: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
 
 env:
-  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
+  # FIXME
+  PUSH_TO_HUB: true  # ${{ github.event_name == 'schedule' }}
```
Comment (Borda) [inline .github/workflows/ci-pytorch-dockers.yml:28]:
```suggestion
  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
```

File: .github/checkgroup.yml [hunk 4]
```diff
@@ -141,15 +137,15 @@ subprojects:
...
+      - "build-cuda (3.10, 1.13, 11.6.1)"
       - "build-hpu (1.5.0, 1.11.0)"
       - "build-ipu (3.9, 1.9)"
       - "build-NGC"
-      - "build-pl (3.9, 1.10, 11.3.1)"
-      - "build-pl (3.9, 1.11, 11.3.1)"
-      - "build-pl (3.9, 1.12, 11.6.1)"
+      - "build-pl (3.10, 1.11, 11.3.1)"
+      - "build-pl (3.10, 1.12, 11.6.1)"
+      - "build-pl (3.10, 1.13, 11.6.1)"
```
Comment (akihironitta) [inline .github/checkgroup.yml:148]: These docker builds don't match with the changes added to the docker workflow.

**Review by akihironitta — COMMENTED** [review]
I wonder if we can let GHA check if a change in `checkgroup.yml` is correct in a PR. It's not feasible to check if there's any discrepancy between `checkgroup.yml` and the CI jobs that are actually triggered every time we change `checkgroup.yml`.

@carmocca Is it possible to run checkgroup with `.github/checkgroup.yml` changed in this branch in addition to the one in master?

**Comment by Borda** [comment]
> I wonder if we can let GHA check if a change in `checkgroup.yml` is correct in a PR. It's not feasible to check if there's any discrepancy between `checkgroup.yml` and the CI jobs that are actually triggered every time we change `checkgroup.yml`.

that would be great to have something like "dry-run"

File: .github/checkgroup.yml (hunk 4 above)
Comment (Borda) [reply .github/checkgroup.yml:148]: Good catch; could you send suggestions...

Comment (akihironitta) [reply .github/checkgroup.yml:149]: Let me directly push a commit to fix this :)

**Review by akihironitta — APPROVED** [review]

File: .github/checkgroup.yml [hunk 5]
```diff
@@ -26,28 +26,28 @@ subprojects:
...
-      - "pl-cpu (windows-2022, lightning, 3.10, 1.12)"
-      #- "pl-cpu (windows-2022, lightning, 3.7, 1.9, oldest)"
+      - "pl-cpu (macOS-11, lightning, 3.10, 1.13)"
+      - "pl-cpu (macOS-11, lightning, 3.7, 1.9, oldest)"
+      - "pl-cpu (ubuntu-20.04, lightning, 3.10, 1.13)"
+      - "pl-cpu (ubuntu-20.04, lightning, 3.7, 1.9, oldest)"
+      - "pl-cpu (windows-2022, lightning, 3.10, 1.13)"
+      - "pl-cpu (windows-2022, lightning, 3.7, 1.9, oldest)"
+      - "make-doctest (pytorch)"
+      - "make-html (pytorch)"
```
Comment (carmocca) [inline .github/checkgroup.yml:50]: This shouldn't be here. Probably from a bad merge

```suggestion
```

File: .github/checkgroup.yml [hunk 6]
```diff
@@ -217,18 +211,15 @@ subprojects:
       - "lite-cpu (macOS-11, lite, 3.9, 1.11)"
       - "lite-cpu (macOS-11, lite, 3.8, 1.10)"
       - "lite-cpu (macOS-11, lite, 3.10, 1.12)"
-      - "lite-cpu (macOS-11, lite, 3.10, 1.13, pre)"
```
Comment (carmocca) [inline .github/checkgroup.yml:220]: Where are the new non-pre 1.13 jobs?

File: requirements/pytorch/check-avail-strategies.py [hunk 7]
```diff
@@ -1,7 +1,17 @@
...
-import deepspeed  # noqa: F401
-import fairscale  # noqa: F401
-import horovod.torch
+import sys
 
-# returns an error code
-assert horovod.torch.nccl_built()
+if __name__ == "__main__":
+    args = sys.argv[1:]
+    print(args)
```
Comment (carmocca) [inline requirements/pytorch/check-avail-strategies.py:5]:
```suggestion
    if not args:
        raise ValueError("Expected at least 1 argument")
```

File: requirements/pytorch/check-avail-strategies.py [hunk 8]
```diff
@@ -1,7 +1,17 @@
...
+
+    if not args or "bagua" in args:
+        import bagua  # noqa: F401
+    if not args or "deepspeed" in args:
+        import deepspeed  # noqa: F401
+    if not args or "fairscale" in args:
+        import fairscale  # noqa: F401
+    if not args or "horovod" in args:
```
Comment (carmocca) [inline requirements/pytorch/check-avail-strategies.py:13]:
```suggestion
    if "bagua" in args:
        import bagua  # noqa: F401
    if "deepspeed" in args:
        import deepspeed  # noqa: F401
    if "fairscale" in args:
        import fairscale  # noqa: F401
    if "horovod" in args:
```

File: requirements/pytorch/check-avail-strategies.py [hunk 9]
```diff
@@ -1,7 +1,17 @@
-import bagua  # noqa: F401
-import deepspeed  # noqa: F401
-import fairscale  # noqa: F401
-import horovod.torch
+import sys
 
-# returns an error code
-assert horovod.torch.nccl_built()
+if __name__ == "__main__":
+    args = sys.argv[1:]
```
Comment (carmocca) [inline requirements/pytorch/check-avail-strategies.py:4]: Where are we making the modification to pass arguments to this? Maybe I'm blind but not seeing it

**Review by carmocca — CHANGES_REQUESTED** [review]
checkgroup needs fixes. I'll check it out when I can!

File: requirements/pytorch/check-avail-strategies.py (hunk 7 above)
Comment (Borda) [reply requirements/pytorch/check-avail-strategies.py:5]: why, by default we would run all

File: requirements/pytorch/check-avail-strategies.py (hunk 9 above)
Comment (Borda) [reply requirements/pytorch/check-avail-strategies.py:4]: it is in building docker images and later on GPU workflows

Comment (Borda) [reply requirements/pytorch/check-avail-strategies.py:4]: well I do not see it either so would need to check what happened

File: .github/checkgroup.yml (hunk 6 above)
Comment (Borda) [reply .github/checkgroup.yml:220]: Good catch, just fixing... but it is example when you update the list only at one place :(

File: .github/workflows/ci-pytorch-dockers.yml [hunk 10]
```diff
@@ -100,6 +102,8 @@ jobs:
           - {python_version: "3.9", pytorch_version: "1.10", cuda_version: "11.3.1"}
           - {python_version: "3.9", pytorch_version: "1.11", cuda_version: "11.3.1"}
           - {python_version: "3.9", pytorch_version: "1.12", cuda_version: "11.6.1"}
+          # TODO: just a few strategies have build support for PT1.13, for example Bagua and ColossalAI are missing
```
Comment (Borda) [inline .github/workflows/ci-pytorch-dockers.yml:105]: @carmocca I have tried to build these images for 1.13 but it failed for missing bagua and colosalAI, so the options are:
- parametrize docker to selectively skip some 3rd part dependencies
- drop problematic packages completely so they won't be tested with GPU as now we run only latest configuration (so we can start building the oldest and newest on GPU)

anyway I would leave it for another PR

File: requirements/pytorch/check-avail-strategies.py (hunk 9 above)
Comment (Borda) [reply requirements/pytorch/check-avail-strategies.py:4]: found it here: https://github.com/Lightning-AI/lightning/pull/15403/files#r1018509556

File: .github/workflows/ci-pytorch-dockers.yml (hunk 10 above)
Comment (carmocca) [reply .github/workflows/ci-pytorch-dockers.yml:105]: Taking care of it in https://github.com/Lightning-AI/lightning/pull/15583. Let's land this one after that one

**Comment by Borda** [comment]
@carmocca, mind checking this again? :rabbit:

**Review by awaelchli — APPROVED** [review]

File: requirements/lite/base.txt [hunk 11]
```diff
@@ -2,7 +2,7 @@
 #  in case you want to preserve/enforce restrictions on the latest compatible version, add "strict" as an in-line comment
 
 numpy>=1.17.2, <1.23.1
-torch>=1.9.*, <=1.13.0
+torch>=1.10.0, <=1.13.0
```
Comment (akihironitta) [inline requirements/lite/base.txt:5]: AFAIK, `.*` was there on purpose so that the CI uses the most recent patch version within the minor version.
```suggestion
torch>=1.10.*, <=1.13.0
```

File: requirements/pytorch/base.txt [hunk 12]
```diff
@@ -2,7 +2,7 @@
 #  in case you want to preserve/enforce restrictions on the latest compatible version, add "strict" as an in-line comment
 
 numpy>=1.17.2, <1.23.1
-torch>=1.9.*, <=1.13.0
+torch>=1.10.0, <=1.13.0
```
Comment (akihironitta) [inline requirements/pytorch/base.txt:5]:
```suggestion
torch>=1.10.*, <=1.13.0
```

File: requirements/lite/base.txt (hunk 11 above)
Comment (Borda) [reply requirements/lite/base.txt:5]: true, let's add it as a comment for our future us :)

**Review by carmocca — APPROVED** [review]
LGTM. Lite and PyTorch CPU configs now match

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
`Test PyTorch / pl-cpu (macOS-11, pytorch, 3.7, 1.10, oldest)` is consistently hanging. We weren't testing this combination before. I suggest we just drop it and move on

**Comment by Borda** [comment]
> I suggest we just drop it and move on.

I ma fine with drop :)

**Comment by lantiga** [comment]
100% drop!
